### PR TITLE
add: dialect sql.Selector LeftJoin

### PR DESCRIPTION
**Pull request by maxiloEmmmm**
### add
- sql.Selector LeftJoin
- sql.Selector RightJoin
- sql.Selector join
- sql.Selector Ons
- sql.EQCol
- sql.NEQCol

### change
- sql.join Struct
- sql.Selector Join
- sql.Selector Clone
- sql.Selector Query
- sql.join

### CODE REVIEW

File: dialect/sql/builder.go [hunk 1]
```diff
@@ -908,6 +914,12 @@ func NEQ(col string, value interface{}) *Predicate {
 	return P().NEQ(col, value)
 }
 
+// EQCol returns a "=" predicate for "col<>col", omit Raw.
+// example: NEQ(`created_at`, Raw(`updated_at`)) <=> NEQCol(`created_at`, `updated_at`)
+func NEQCol(col string, col2 string) *Predicate {
+	return P().NEQ(col, Raw(col2))
+}
+
```
Comment (a8m) [inline dialect/sql/builder.go:922]: This 2 functions are not really needed by the default codegen, and can also be written as follows:

```go
p := sql.P(func(b *sql.Builder) {
	b.Ident("a").WriteOp(sql.OpEQ).Ident("b") // OpEQ can be replaced with any other operator.
})
```
Let's remove them.

File: dialect/sql/builder.go [hunk 2]
```diff
@@ -1423,7 +1435,7 @@ func (*SelectTable) view() {}
 
 // join table option.
 type join struct {
-	on    string
+	on    *Predicate
```
Comment (a8m) [inline dialect/sql/builder.go:1438]: 👍

File: dialect/sql/builder.go [hunk 3]
```diff
@@ -1563,8 +1576,23 @@ func (s *Selector) Table() *SelectTable {
...
+	return s.join("LEFT JOIN", t)
+}
+
+// RightJoin appends a `RIGHT JOIN` clause to the statement.
+func (s *Selector) RightJoin(t TableView) *Selector {
+	return s.join("RIGHT JOIN", t)
+}
+
+// join
+func (s *Selector) join(kind string, t TableView) *Selector {
```
Comment (a8m) [inline dialect/sql/builder.go:1593]: Improve comment please.

File: dialect/sql/builder.go [hunk 4]
```diff
@@ -1604,14 +1632,29 @@ func (s *Selector) Columns(columns ...string) []string {
...
 	if len(s.joins) > 0 {
-		s.joins[len(s.joins)-1].on = fmt.Sprintf("%s = %s", c1, c2)
+		join := &s.joins[len(s.joins)-1]
+
+		switch {
+		case join.on == nil:
+			join.on = p
+		default:
+			join.on = And(join.on, p)
+		}
```
Comment (a8m) [inline dialect/sql/builder.go:1645]: Change it to `OnP`.

File: dialect/sql/builder.go [hunk 5]
```diff
@@ -1604,14 +1632,29 @@ func (s *Selector) Columns(columns ...string) []string {
...
 	return s
 }
 
+// On sets the `ON` clause for the `JOIN` operation.
+func (s *Selector) On(c1, c2 string) *Selector {
+	s.Ons(P().Append(func(builder *Builder) {
+		builder.WriteString(c1).WriteOp(OpEQ).WriteString(c2)
+	}))
+	return s
+}
```
Comment (a8m) [inline dialect/sql/builder.go:1656]: Also fix comment please. 
```suggestion
func (s *Selector) On(c1, c2 string) *Selector {
	s.OnP(P(func(b *Builder) {
		b.Ident(c1).WriteOp(OpEQ).Ident(c2)
	}))
	return s
}
```

**Review by a8m — CHANGES_REQUESTED** [review]
Thanks for your contribution @maxiloEmmmm! It looks good.

Minor comments.

File: dialect/sql/builder.go (hunk 1 above)
Comment (maxiloEmmmm) [reply dialect/sql/builder.go:922]: I might just want to save two lines of code
will remove *EQCol

File: dialect/sql/builder.go (hunk 3 above)
Comment (maxiloEmmmm) [reply dialect/sql/builder.go:1593]: *Join(Join, LeftJoin....) use it together

File: dialect/sql/builder.go (hunk 4 above)
Comment (maxiloEmmmm) [reply dialect/sql/builder.go:1645]: if use `Onp`
switch case default `And` seems redundant
i think so
remove default case ?

File: dialect/sql/builder.go (hunk 3 above)
Comment (maxiloEmmmm) [reply dialect/sql/builder.go:1581]: because: https://github.com/facebook/ent/issues/805

File: dialect/sql/builder.go (hunk 4 above)
Comment (a8m) [reply dialect/sql/builder.go]: I meant, change the function name from `Ons` to `OnP` (i.e. "On Predicate").

File: dialect/sql/builder.go [hunk 6]
```diff
@@ -1604,14 +1620,22 @@ func (s *Selector) Columns(columns ...string) []string {
 	return names
 }
 
-// On sets the `ON` clause for the `JOIN` operation.
-func (s *Selector) On(c1, c2 string) *Selector {
+// Ons sets or appends the given predicate to the statement.
```
Comment (a8m) [inline dialect/sql/builder.go:1623]:
```suggestion
// OnP sets or appends the given predicate for the `ON` clause of the statement.
```

Comment fix.

File: dialect/sql/builder.go [hunk 7]
```diff
@@ -1563,8 +1564,23 @@ func (s *Selector) Table() *SelectTable {
...
+	return s.join("LEFT JOIN", t)
+}
+
+// RightJoin appends a `RIGHT JOIN` clause to the statement.
+func (s *Selector) RightJoin(t TableView) *Selector {
+	return s.join("RIGHT JOIN", t)
+}
+
+// join
+func (s *Selector) join(kind string, t TableView) *Selector {
```
Comment (a8m) [inline dialect/sql/builder.go:1581]:
```suggestion
// join adds a join table to the selector with the given kind. 
func (s *Selector) join(kind string, t TableView) *Selector {
```

**Review by a8m — APPROVED** [review]
2 last comments. Let's fix them and I'll land it.

Thanks @maxiloEmmmm

File: dialect/sql/builder.go (hunk 7 above)
Comment (maxiloEmmmm) [reply dialect/sql/builder.go:1581]: Thank you for your help

**Pull request merged**